rastertosag-gdi (cups driver): init at 0.1

Yarny0 opened the pull request:
###### Motivation for this change

Some Ricoh printers use the proprietary sag-gdi format and can't handle other, more common formats.  This commit brings a filter for cups that generates the sag-gdi format.

The latest version 0.1 is dated 2011.  So updates are unlikely.

The filter is written for Python 2.  To avoid new reverse dependencies on Python 2, the commit employs a patch from Debian that migrates the code to Python 3.

The README file just states "GPL" as license.  It is unclear whether that refers to the first version or to the "current version" in the year of the copyright (would be 3), and whether newer versions would be included.  The commit picks the nixpkgs `free` license as this seems to be the most general license covering all possible GPL combinations.  At least, `free` should permit Hydra to build the package.

The source tarball brings pdd files, but also a drv file that can be used to generate those ppd files.  Some discussions on Discource left me with the impression that building from source is generally preferable over using prebuilt files.  Hence the commit calls cups' `ppdc` to build ppd files from the drv file.

Here is a documentation of the sag-gdi format (with expired TLS cert, unfortunatelly): https://www.undocprint.org/formats/page_description_languages/sagem-gdi

There is no namespace for cups drivers (unlike the `xorg` attribute set, as far as I can tell), so I guess it's a good idea to prefix the package's attribute name with `cups-drv-`.


###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS: x86-64
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

More:

Tested successfully with a Philips 6050W 4-in-1 printer (connected via USB, with ppd file `rsp1000s.ppd`).  The printer identifes as `Sagem` in the USB device listing.  While the computer that is connected to the printer runs NixOS 20.09, for this test, the driver itself was build with the nixpkgs version the commit is based on (nixos-unstable commit ad47284f8b01f587e24a4f14e0f93126d8ebecda).  So the Python interpreter for the script and the `ppdc` binary that created the ppd files for this test all come from a recent nixos-unstable branch.

Yarny0 committed:
rastertosag-gdi (cups driver): init at 0.1

Some Ricoh printers use the proprietary sag-gdi format
and can't handle other, more common formats.
This commit brings a filter for cups
that generates the sag-gdi format.

The latest version 0.1 is dated 2011.  So updates are unlikely.

The filter is written for Python 2.
To avoid new reverse dependencies on Python 2,
we employ a patch from Debian
that migrates the code to Python 3.

The README file just states "GPL" as license.
It is unclear whether that refers to the first version or
to the "current version" in the year of the copyright
(would be 3), and whether newer versions would be included.
The commit picks the nixpkgs `free` license
as this seems to be the most general license
covering all possible GPL combinations.
At least, `free` should permit Hydra to build the package.

The source tarball brings pdd files, but also a
drv file that can be used to generate those ppd files.
Since we prefer building from source,
we call cups' `ppdc` to build ppd files from the drv file.

Here is a documentation of the sag-gdi format:
https://www.undocprint.org/formats/page_description_languages/sagem-gdi

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-drv-rastertosag-gdi/default.nix:
```diff
@@ -0,0 +1,60 @@
+    })
+  ];
+  format = "other";
+  nativeBuildInputs = [ (lib.attrsets.getBin cups) ];
```
```suggestion
  nativeBuildInputs = [ (lib.getBin cups) ];
```

Yarny0 replied:
Thanks for the hint and commit.  That's fine with me.  I usually use subsets of `lib` as I feel these are more readable and more self-documenting.  But it doesn't really make a difference here.

SuperSandro2000 reviewed (commented):
Nix wise this is fine except that we usually use the shorthands for ln and install but I have no knowledge about cups driver.

Yarny0 commented:
Thanks for your hints, @SuperSandro2000 .  I updated the branch, included your shortening of `lib.attrs`, and replaced the command line arguments with short versions for uniformity.

Maybe it's a good idea to notify other cups driver maintainers for a review?  @steveej and @makefu , you're maintaining two cups driver packages (in `/pkgs/misc/cups/drivers`).  May I ask you for a review, if you can spare the time?  That would be very kind.

makefu reviewed (approved):
It's been a long time since i've packaged my cups driver. The package looks very clean, haven't tested it though due to the lack of this specific hardware :)
